rclpy/src/rclpy/serialization.cpp

```diff
@@ -0,0 +1,149 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Include pybind11 before rclpy_common/handle.h includes Python.h
+#include <pybind11/pybind11.h>
+
+#include <rcl/error_handling.h>
+#include <rcl/rcl.h>
+#include <rcl/types.h>
+#include <rcutils/allocator.h>
+#include <rcutils/error_handling.h>
+
+#include <memory>
+#include <stdexcept>
+
+extern "C"
+{
+#include "rclpy_common/common.h"
+}
+#include "rclpy_common/handle.h"
+
+#include "rclpy_common/exceptions.hpp"
+
+#include "serialization.hpp"
+
+namespace rclpy
+{
+class SerializedMessage
+{
+public:
+  SerializedMessage()
+  {
+    rcl_msg_ = rmw_get_zero_initialized_serialized_message();
+  }
+
+  void
+  init(rcutils_allocator_t allocator)
+  {
+    rcutils_ret_t rcutils_ret = rmw_serialized_message_init(&rcl_msg_, 0u, &allocator);
+    if (RCUTILS_RET_OK != rcutils_ret) {
+      throw RCLError("failed to initialize serialized message");
+    }
+  }
+
+  ~SerializedMessage()
+  {
+    rcutils_ret_t ret = rmw_serialized_message_fini(&rcl_msg_);
+    if (RCUTILS_RET_OK != ret) {
+      fprintf(
+        stderr,
+        "[rclpy|" RCUTILS_STRINGIFY(__FILE__) ":" RCUTILS_STRINGIFY(__LINE__) "]: "
+        "failed to fini rcl_serialized_msg_t (%d) in destructor: %s\n",
+        ret,
+        rcutils_get_error_string().str);
+      rcutils_reset_error();
+    }
+  }
+
+  rcl_serialized_message_t rcl_msg_;
+};
+
+py::bytes
+serialize(py::object pymsg, py::object pymsg_type)
+{
+  // Get type support
+  auto ts = static_cast<rosidl_message_type_support_t *>(
+    rclpy_common_get_type_support(pymsg_type.ptr()));
+  if (!ts) {
+    throw py::error_already_set();
+  }
+
+  destroy_ros_message_signature * destroy_ros_message = NULL;
+  auto deleter = [&](void * ptr) {destroy_ros_message(ptr);};
+  auto ros_msg = std::unique_ptr<void, decltype(deleter)>(
+    rclpy_convert_from_py(pymsg.ptr(), &destroy_ros_message), deleter);
+  if (!ros_msg) {
+    throw py::error_already_set();
+  }
+
+  // Create a serialized message object
+  SerializedMessage serialized_msg;
+  serialized_msg.init(rcutils_get_default_allocator());
+
+  // Serialize
+  rmw_ret_t rmw_ret = rmw_serialize(ros_msg.get(), ts, &serialized_msg.rcl_msg_);
+  if (RMW_RET_OK != rmw_ret) {
+    throw RCLError("Failed to serialize ROS message");
+  }
```

Comment:
This isn't the right way to handle RMW errors. There are explicit functions defined for it, but they're all just typedefs to `rcutils` error handling. It would be OK with me if you just switched this to `rcutils` error handling, but I think the more correct way to do it would be to introduce an `RMWError` exception class like you did for `RCUtilsError`.

Reply:
Fully agreed.

Reply:
I'll update this to use RMWError from #709 once that's merged.

Reply:
Uses RMWError in 5204de18f926f1586f37e248f572cc5d4455416f